contents/articles/2020-11-29-creer-plugin-eslint-typescript-magie-ast.md

```diff
@@ -0,0 +1,476 @@
+---
+date: 2020-11-29
+title: "Créer votre propre règle ESLint pour TypeScript: la magie des AST"
+author: gomesalexandre
+slug: creer-plugin-eslint-typescript-magie-ast
+---
+Aujourd'hui, découvrons comment créer un plugin ESLint pour écrire du code à notre place, et automatiser ce qui prendrait des semaines, voire des mois manuellement.
+
+
+![Le résultat? Une règle TypeScript ESLint qui ajoutera automatiquement le génerique nécessaire à vos appels vers shallow(), pour ne pas avoir d'erreurs de typage.](/public/images/articles/2020-11-29-creer-plugin-eslint-typescript-magie-ast/result-gif.gif)*Le résultat? Une règle TypeScript ESLint qui ajoutera automatiquement le génerique nécessaire à vos appels vers shallow(), pour ne pas avoir d'erreurs de typage.*
+
+
+Découvrez le monde merveilleux des "AST": ils ne sont pas si difficiles que ce que l'on pourrait croire!
+
+# Pourquoi écrire vos propres règles ESLint ?
+
+* C'est intéressant, et une bonne manière d'approfondir vos connaissances syntactiques à propos de JS/TS 
+
+* Cela vous permet de définir des règles quant au style du code au sein de votre entreprise
+
+* Ce sont potentiellement des semaines de travail "manuel" en moins 😃
+
+Il existe déjà plein de règles, dictant [le style de vos accolades](https://eslint.org/docs/rules/brace-style), le fait [que vos fonctions async de ne doivent pas retourner d'expression await](https://eslint.org/docs/rules/no-return-await) ou encore [le fait que vous ne devez pas initialiser vos variables avec la valeur `undefined`](https://eslint.org/docs/rules/no-undef-init).
```

Comment:
Comme dis plus haut, je trouve le premier exemple inutile (et même contre productif) et le dernier pas spécialement pertinent: en quoi c'est bien d'interdire undefined? Ca mérite un débat presque... (perso j'utilise undefined pour simuler des valeurs optionnels en JavaScript, et à par contre je n'utilise pas null)

Reply:
Ce n'est pas "bien" ni son contraire 😄  l'idée est de rappeler l'étendue des règles eslint présentes par défaut (sans compter les plugins comme react, node etc)
La deuxième n'est d'ailleurs pas [fondalement nécessaire](https://twitter.com/addaleax/status/1323966960951042049) non plus (Anna fait partie de la core team Node)

eslint et prettier ont deux philosophies différentes, l'un est "batteries included" avec des opinions fortes et un tas de conventions, l'autre propose de créer ses conventions avec de la granularité quant aux règles
Que l'on trouve de la valeur ou pas à certaines règles (je suis le premier à raler sur le choix de certaines règles dans des projets), c'est bien pour cette flexibilité qu'on utilise eslint. 

Fondamentalement, prettier, c'est juste un ensemble de règles, il est d'ailleurs souvent utilisé en conjonction avec eslint (`eslint-plugin-prettier` lance prettier comme si c'était une règle eslint)

Reply:
Je connais bien eslint est son approche, je trouve juste ca complètement inutile depuis que prettier existe du pourquoi de on devrait caller un espace avec une accolage (ou pas) etc. Sans *aucune* règle, forcément c'est l'anarchie, mais avec un "pseudo" standard comme prettier, utiliser par des millions de développeurs (and counting), je ne vois vraiment aucun intérêt à eslint pour le style. Je pense qu'il n'y a pas de débat à avoir à ce niveau. Des gens voudront se battre pour faire "tabs v space" ou "comma or not comma" mais au quotidien, ca reste des efforts qui n'apportent vraiment rien. Je m'égare mais au final je pense qu'on sera tous d'accord dans le crew putaindecode pour dire que parler d'eslint pour du style n'est clairement pas à son avantage à ce jour.

Reply:
Tu prèches un convaincu, je prèfère prettier dans 99% des cas, j'adhère en majorité aux opinions qu'ils ont choisi 😉  et dans le cas contraire, comme tu l'as dit c'est le standard un peu partout (je suis en ce moment-même sur un projet sur Codesandbox, où il est activé par défaut pour tout le monde).
Du coup la tournure est changée.